### PR TITLE
perf(ext/websocket): replace tokio_tungstenite server with fastwebsockets

### PR DESCRIPTION
**Pull request by littledivy**
https://github.com/littledivy/fastwebsockets

```
# This PR
./load_test 100 0.0.0.0 8080 0 0
Running benchmark now...
Msg/sec: 176355.000000

# main
./load_test 100 0.0.0.0 8080 0 0
Running benchmark now...
Msg/sec: 157198.750000
```

### CODE REVIEW

File: profile.json
Comment (bartlomieju) [inline profile.json:1]: Remove this file

File: ext/websocket/server.rs [hunk 1]
```diff
@@ -0,0 +1,191 @@
...
+        MessageKind::Error as u16,
+        StringOrBuffer::String(err.to_string()),
+      ))
+    }
+  };
+
+  let res = match val.opcode {
+    OpCode::Text => (
+      MessageKind::Text as u16,
+      StringOrBuffer::String(String::from_utf8(val.payload).unwrap()),
```
Comment (mmastrac) [inline ext/websocket/server.rs:162]: Should this return an error instead of `panic`ing?

File: ext/websocket/server.rs [hunk 2]
```diff
@@ -0,0 +1,191 @@
...
+  state: Rc<RefCell<OpState>>,
+  rid: ResourceId,
+  code: Option<u16>,
+  reason: Option<String>,
+) -> Result<(), AnyError> {
+  let resource = state
+    .borrow_mut()
+    .resource_table
+    .get::<ServerWebSocket>(rid)?;
+  let mut ws = RcRef::map(&resource, |r| &r.ws).borrow_mut().await;
```
Comment (mmastrac) [inline ext/websocket/server.rs:129]: If (code, reason) are (None, None), this should probably return (1005, "").

**Review by mmastrac — COMMENTED** [review]
Nice! LGTM w/two small comments. Letting @bartlomieju give the final stamp.

File: ext/websocket/server.rs (hunk 1 above)
Comment (littledivy) [reply ext/websocket/server.rs:162]: fastws guarantees Text payloads are always utf8. Its not obvious yet in the documentation, when thats done we can switch this to `String::from_utf8_unchecked`

File: ext/websocket/server.rs (hunk 2 above)
Comment (littledivy) [reply ext/websocket/server.rs:129]: Body in close frame is optional.

https://www.rfc-editor.org/rfc/rfc6455#section-5.5.1

Section 5.5.1:

> The Close frame MAY contain a body (the "Application data" portion of
   the frame) that indicates a reason for closing

ref: tungstenite https://github.com/snapview/tungstenite-rs/blob/92d65e11043f2a3bf18099ecc3cd7911e70bc456/src/protocol/frame/frame.rs#L347

https://www.rfc-editor.org/rfc/rfc6455#section-7.4

Section 7.4:

> 1005 is a reserved value and MUST NOT be set as a status code in a
      Close control frame by an endpoint.  It is designated for use in
      applications expecting a status code to indicate that no status
      code was actually present.

File: ext/websocket/01_websocket.js [hunk 3]
```diff
@@ -346,7 +360,11 @@ class WebSocket extends EventTarget {
       const d = core.encode(string);
       this[_bufferedAmount] += TypedArrayPrototypeGetByteLength(d);
       PromisePrototypeThen(
-        core.opAsync2("op_ws_send_text", this[_rid], string),
+        core.opAsync2(
+          this[_role] == SERVER ? "op_server_ws_send_text" : "op_ws_send_text",
```
Comment (petamoriken) [inline ext/websocket/01_websocket.js:364]: `===` is better than `==` for performance :)

File: ext/websocket/01_websocket.js [hunk 4]
```diff
@@ -401,7 +419,12 @@ class WebSocket extends EventTarget {
       this[_readyState] = CLOSING;
 
       PromisePrototypeCatch(
-        core.opAsync("op_ws_close", this[_rid], code, reason),
+        core.opAsync(
+          this[_role] == SERVER ? "op_server_ws_close" : "op_ws_close",
```
Comment (petamoriken) [inline ext/websocket/01_websocket.js:423]: ditto

File: ext/websocket/01_websocket.js [hunk 5]
```diff
@@ -422,7 +445,7 @@ class WebSocket extends EventTarget {
   async [_eventLoop]() {
     while (this[_readyState] !== CLOSED) {
       const { 0: kind, 1: value } = await core.opAsync2(
-        "op_ws_next_event",
+        this[_role] == SERVER ? "op_server_ws_next_event" : "op_ws_next_event",
```
Comment (petamoriken) [inline ext/websocket/01_websocket.js:448]: ditto

File: ext/websocket/01_websocket.js [hunk 6]
```diff
@@ -489,7 +512,7 @@ class WebSocket extends EventTarget {
           if (prevState === OPEN) {
             try {
               await core.opAsync(
-                "op_ws_close",
+                this[_role] == SERVER ? "op_server_ws_close" : "op_ws_close",
```
Comment (petamoriken) [inline ext/websocket/01_websocket.js:515]: ditto

File: ext/websocket/01_websocket.js [hunk 7]
```diff
@@ -517,14 +540,23 @@ class WebSocket extends EventTarget {
       clearTimeout(this[_idleTimeoutTimeout]);
       this[_idleTimeoutTimeout] = setTimeout(async () => {
         if (this[_readyState] === OPEN) {
-          await core.opAsync("op_ws_send", this[_rid], {
-            kind: "ping",
-          });
+          await core.opAsync(
+            this[_role] == SERVER ? "op_server_ws_send" : "op_ws_send",
```
Comment (petamoriken) [inline ext/websocket/01_websocket.js:544]: ditto

File: ext/websocket/01_websocket.js [hunk 8]
```diff
@@ -517,14 +540,23 @@ class WebSocket extends EventTarget {
...
+              kind: "ping",
+            },
+          );
           this[_idleTimeoutTimeout] = setTimeout(async () => {
             if (this[_readyState] === OPEN) {
               this[_readyState] = CLOSING;
               const reason = "No response from ping frame.";
-              await core.opAsync("op_ws_close", this[_rid], 1001, reason);
+              await core.opAsync(
+                this[_role] == SERVER ? "op_server_ws_close" : "op_ws_close",
```
Comment (petamoriken) [inline ext/websocket/01_websocket.js:555]: ditto

**Review by bartlomieju — APPROVED** [review]
LGTM, curious to see benchmark results on `main`

**Comment by github-enetrino-com** [comment]
this commit breaks websockets

sends do not go through until new messages arrive

it's easy to reproduce with example close to the one from https://medium.com/deno-the-complete-reference/native-web-sockets-client-server-in-deno-928678a65cf2 replacing prompts to input text with setTimeout with sending ping-pongs

```
//chat_server.ts

import { serve } from "https://deno.land/std/http/mod.ts";
function logError(msg: string) {
  console.log(msg);
  Deno.exit(1);
}
function handleConnected() {
  console.log("Connected to client ...");
}
function handleMessage(ws: WebSocket, data: string) {
  console.log("CLIENT >> " + data);
  setTimeout(()=>ws.send('pong'),2000);
  // const reply = prompt("Server >> ") || "No reply";
  // if (reply === "exit") {
  //   return ws.close();
  // }
  // ws.send(reply as string);
}
function handleError(e: Event | ErrorEvent) {
  console.log(e instanceof ErrorEvent ? e.message : e.type);
}
async function reqHandler(req: Request) {
  if (req.headers.get("upgrade") != "websocket") {
    return new Response(null, { status: 501 });
  }
  const { socket: ws, response } = Deno.upgradeWebSocket(req);
  ws.onopen = () => handleConnected();
  ws.onmessage = (m) => handleMessage(ws, m.data);
  ws.onclose = () => logError("Disconnected from client ...");
  ws.onerror = (e) => handleError(e);
  return response;
}
console.log("Waiting for client ...");
serve(reqHandler, { port: 8001 });
```

```
//chat_client.ts

function logError(msg: string) {
  console.log(msg);
  Deno.exit(1);
}
function handleConnected(ws: WebSocket) {
  console.log("Connected to server ...");
  handleMessage(ws, "Welcome!");
}
function handleMessage(ws: WebSocket, data: string) {
  console.log("SERVER >> " + data);
  setTimeout(()=>ws.send('ping'),2000);
  // const reply = prompt("Client >> ") || "No reply";
  // if (reply === "exit") {
  //   return ws.close();
  // }
  // ws.send(reply as string);
}
function handleError(e: Event | ErrorEvent) {
  console.log(e instanceof ErrorEvent ? e.message : e.type);
}
console.log("Connecting to server ...");
try {
  const ws = new WebSocket("ws://localhost:8001");
  ws.onopen = () => handleConnected(ws);
  ws.onmessage = (m) => handleMessage(ws, m.data);
  ws.onclose = () => logError("Disconnected from server ...");
  ws.onerror = (e) => handleError(e);
} catch (err) {
  logError("Failed to connect to server ... exiting");
}
```

this example sends ping-pongs every 2 seconds on 1.32.3, but stalls on 1.32.4

**Pull request merged**